lib/cancan/model_adapters/active_record_adapter.rb

```diff
@@ -3,68 +3,40 @@
 module CanCan
   module ModelAdapters
     class ActiveRecordAdapter < AbstractAdapter
-      def self.version_greater_or_equal?(version)
-        Gem::Version.new(ActiveRecord.version).release >= Gem::Version.new(version)
+      def self.for_class?(klass)
+        klass < ActiveRecord::Base
       end
 
-      def self.version_lower?(version)
-        Gem::Version.new(ActiveRecord.version).release < Gem::Version.new(version)
-      end
-
-      def initialize(model_class, rules)
+      def initialize(relation_or_class, rules)
         super
+        @model_class = relation_or_class.respond_to?(:klass) ? relation_or_class.klass : relation_or_class
+        @relation = relation_or_class.all
         @compressed_rules = RulesCompressor.new(@rules.reverse).rules_collapsed.reverse
         StiNormalizer.normalize(@compressed_rules)
-        ConditionsNormalizer.normalize(model_class, @compressed_rules)
-      end
-
-      # Returns conditions intended to be used inside a database query. Normally you will not call this
-      # method directly, but instead go through ModelAdditions#accessible_by.
-      #
-      # If there is only one "can" definition, a hash of conditions will be returned matching the one defined.
-      #
-      #   can :manage, User, :id => 1
-      #   query(:manage, User).conditions # => { :id => 1 }
-      #
-      # If there are multiple "can" definitions, a SQL string will be returned to handle complex cases.
-      #
-      #   can :manage, User, :id => 1
-      #   can :manage, User, :manager_id => 1
-      #   cannot :manage, User, :self_managed => true
-      #   query(:manage, User).conditions # => "not (self_managed = 't') AND ((manager_id = 1) OR (id = 1))"
-      #
-      def conditions
-        conditions_extractor = ConditionsExtractor.new(@model_class)
-        if @compressed_rules.size == 1 && @compressed_rules.first.base_behavior
-          # Return the conditions directly if there's just one definition
-          conditions_extractor.tableize_conditions(@compressed_rules.first.conditions).dup
-        else
-          extract_multiple_conditions(conditions_extractor, @compressed_rules)
-        end
-      end
-
-      def extract_multiple_conditions(conditions_extractor, rules)
-        rules.reverse.inject(false_sql) do |sql, rule|
-          merge_conditions(sql, conditions_extractor.tableize_conditions(rule.conditions).dup, rule.base_behavior)
+        ConditionsNormalizer.normalize(@model_class, @compressed_rules)
+        # run the extractor on the reversed set of rules to fill the cache properly
+        @conditions_extractor = ConditionsExtractor.new(@model_class).tap do |extractor|
+          @compressed_rules.reverse_each { |rule| extractor.tableize_conditions(rule.conditions) }
```

Comment:
I wonder if these `reverse`'s are necessary (line 14, then re-reversed on line 19)

Reply:
They are, because the `RulesCompressor` has an implicit requirement on the ordering of the conditions, unfortunately.